// renderkid/scripts/js/lib/RenderKid.js
var AnsiPainter, Layout, RenderKid, Styles, blockStyleApplier, inlineStyleApplier, object, renderKidConfig, tools;

inlineStyleApplier = require('./renderKid/styleApplier/inline');

blockStyleApplier = require('./renderKid/styleApplier/block');

renderKidConfig = require('./renderKid/config');

AnsiPainter = require('./AnsiPainter');

Layout = require('./Layout');

Styles = require('./renderKid/Styles');

tools = require('./tools');

object = require('utila').object;

module.exports = RenderKid = (function() {
  var self;

  self = RenderKid;

  RenderKid.AnsiPainter = AnsiPainter;

  RenderKid.Layout = Layout;

  RenderKid._defaultConfig = {
    layout: {
      terminalWidth: 80
    }
  };

  RenderKid.quote = tools.quote;

  function RenderKid(config) {
    if (config == null) {
      config = {};
    }
    this._config = renderKidConfig(config, self._defaultConfig);
    this._initStyles();
  }

  RenderKid.prototype._initStyles = function() {
    return this._styles = new Styles;
  };

  RenderKid.prototype.style = function() {
    return this._styles.setRule.apply(this._styles, arguments);
  };

  RenderKid.prototype._getStyleFor = function(el) {
    return this._styles.getStyleFor(el);
  };

  RenderKid.prototype.render = function(input) {
    return this._paint(this._renderDom(this._toDom(input)));
  };

  RenderKid.prototype._toDom = function(input) {
    if (typeof input === 'string') {
      return this._parse(input);
    } else if (object.isBareObject(input) || Array.isArray(input)) {
      return this._objToDom(input);
    } else {
      throw Error("Invalid input type. Only strings, arrays and objects are accepted");
    }
  };

  RenderKid.prototype._objToDom = function(o, injectFakeRoot) {
    if (injectFakeRoot == null) {
      injectFakeRoot = true;
    }
    if (injectFakeRoot) {
      o = {
        body: o
      };
    }
    return tools.objectToDom(o);
  };

  RenderKid.prototype._paint = function(text) {
    return AnsiPainter.paint(text);
  };

  RenderKid.prototype._parse = function(string, injectFakeRoot) {
    if (injectFakeRoot == null) {
      injectFakeRoot = true;
    }
    if (injectFakeRoot) {
      string = '<body>' + string + '</body>';
    }
    return tools.stringToDom(string);
  };

  RenderKid.prototype._renderDom = function(dom) {
    var bodyTag, layout, rootBlock;
    bodyTag = dom[0];
    layout = new Layout(this._config.layout);
    rootBlock = layout.getRootBlock();
    this._renderBlockNode(bodyTag, rootBlock);
    return layout.get();
  };

  RenderKid.prototype._renderChildren = function(nodes, parentBlock) {
    var node, _i, _len;
    for (_i = 0, _len = nodes.length; _i < _len; _i++) {
      node = nodes[_i];
      this._renderNode(node, parentBlock);
    }
  };

  RenderKid.prototype._renderNode = function(node, parentBlock) {
    if (node.type === 'text') {
      this._renderText(node, parentBlock);
    } else if (node.name === 'br') {
      this._renderBr(node, parentBlock);
    } else if (this._isBlock(node)) {
      this._renderBlockNode(node, parentBlock);
    } else {
      this._renderInlineNode(node, parentBlock);
    }
  };

  RenderKid.prototype._renderText = function(node, parentBlock) {
    var text;
    text = node.data;
    text = text.replace(/[\s]+/g, ' ');
    text = text.trim();
    if (text.length === 0) {
      return;
    }
    return parentBlock.write(text);
  };

  RenderKid.prototype._renderBlockNode = function(node, parentBlock) {
    var after, before, block, blockConfig, _ref;
    _ref = blockStyleApplier.applyTo(node, this._getStyleFor(node)), before = _ref.before, after = _ref.after, blockConfig = _ref.blockConfig;
    block = parentBlock.openBlock(blockConfig);
    if (before !== '') {
      block.write(before);
    }
    this._renderChildren(node.children, block);
    if (after !== '') {
      block.write(after);
    }
    return block.close();
  };

  RenderKid.prototype._renderInlineNode = function(node, parentBlock) {
    var after, before, _ref;
    _ref = inlineStyleApplier.applyTo(node, this._getStyleFor(node)), before = _ref.before, after = _ref.after;
    if (before !== '') {
      parentBlock.write(before);
    }
    this._renderChildren(node.children, parentBlock);
    if (after !== '') {
      return parentBlock.write(after);
    }
  };

  RenderKid.prototype._renderBr = function(node, parentBlock) {
    return parentBlock.write("\n");
  };

  RenderKid.prototype._isBlock = function(node) {
    if (node.type === 'text' || node.name === 'br' || this._getStyleFor(node).display !== 'block') {
      return false;
    }
    return true;
  };

  return RenderKid;

})();
